packages/decoupled/src/config/config-validator.ts

```diff
@@ -0,0 +1,157 @@
+
+    constructor(
+        private readonly validatorOptions: Options = {},
+        private readonly TJSSettings: TJS.PartialArgs = {
+            id: 'decoupled.contrib.json',
+            defaultProps: true,
+            noExtraProps: true,
+            required: true,
+        },
+        private readonly TJSCompilerOptions: TJS.CompilerOptions = {
+            strictNullChecks: true,
+            useTypeOfKeyword: true,
+        },
+    ) {
+        this.validator = new Ajv(validatorOptions);
+
+        addKeywords(this.validator, ['typeof']);
+
+        this.contributes = this.loadContributes();
+        this.defsSchema = this.generateDefinitionsSchema();
+        this.schema = this.generateSchemaFromDefinitions();
+    }
+
+    get autoloadPattern() {
+        return `'**/${this.contributeFileName}'`;
+    }
+
+    /**
+     * Validate input data
+     * @param data
+     */
+    public validate(data: AnyObject): [ null | ErrorObject[] , boolean] {
+        const { defsSchema, schema } = this;
+
+        const validate = this.validator
+            .addSchema(defsSchema)
+            .compile(schema);
+        const valid = validate(data);
+
+        return [validate.errors, valid];
+    }
+
+    /**
+     * Generate definitions schema from decoupled contributes
+     */
+    private generateDefinitionsSchema(): DefsSchemaInterface {
+        const { contributes, TJSSettings, TJSCompilerOptions } = this;
+
+        const program = TJS.getProgramFromFiles(contributes, TJSCompilerOptions);
+
+        return TJS.generateSchema(program, '*', TJSSettings, contributes);
+    }
+
+    /**
+     * Generate actual schema from definitions schema
+     */
+    private generateSchemaFromDefinitions(): JSONSchemaInterface {
+        const { defsSchema: { definitions = {} } } = this;
+
+        const properties = Object.entries(definitions).reduce((init, [k, v]) => {
+            init[k.toLowerCase()] = { $ref: `decoupled.contrib.json#/definitions/${k}` };
+            return init;
+        }, {});
+
+        const required = Object.keys(properties)
+            .filter((k) => k[0] !== '$'); // Filter out common types with leading $
+
+        return {
+            type: 'object',
+            properties,
+            required,
+        };
+    }
+
+    /**
+     * Load contributes file locations
+     */
+    private loadContributes() {
+        const modules = getFromDecoupledConfig('modules');
+
+        if (modules && Array.isArray(modules) && modules.length > 0) {
+            return this.loadContributesFromModules(modules);
+        }
+
+        logger.warn('No modules config found in Decoupled Config.');
+
+        return this.autoLoadContributes();
+    }
+
+    /**
+     * Load contributes file locations from decoupled.config
+     *
+     * 1. Root and `src` directories
+     * 2. Sites directories
+     * 3. Modules directories under `node_modules`
+     */
+    private loadContributesFromModules(modules: string[]) {
```

Comment:
naming: this function seems not to load anything, but got find module paths instead